File: resources/recipes/authentication/password-hashing/hash_pbkdf2.js
const crypto = require('crypto');

const ITERATIONS = 600_000;
const KEYLEN = 32;
const DIGEST = 'sha256';

function hashPassword(password) {
  const salt = crypto.randomBytes(16);
  const key = crypto.pbkdf2Sync(password, salt, ITERATIONS, KEYLEN, DIGEST);
  return `pbkdf2_sha256$${ITERATIONS}$${salt.toString('hex')}$${key.toString('hex')}`;
}

function verifyPassword(stored, password) {
  const [_, iters, saltHex, hashHex] = stored.split('$');
  const salt = Buffer.from(saltHex, 'hex');
  const expected = Buffer.from(hashHex, 'hex');
  const derived = crypto.pbkdf2Sync(password, salt, parseInt(iters, 10), expected.length, DIGEST);
  return crypto.timingSafeEqual(derived, expected);
}

const password = 'supersecret';
const start = Date.now();
const stored = hashPassword(password);
const hashMs = Date.now() - start;
console.log(`PBKDF2 stored: ${stored}`);
console.log(`hash time: ${hashMs} ms`);
console.log(`verify: ${verifyPassword(stored, password)}`);

File: resources/recipes/authentication/password-hashing/hash_pbkdf2.py
import hashlib
import hmac
import secrets
import time


def hash_password(password: bytes, iterations: int = 600_000) -> str:
    salt = secrets.token_bytes(16)
    key = hashlib.pbkdf2_hmac("sha256", password, salt, iterations, dklen=32)
    return f"pbkdf2_sha256${iterations}${salt.hex()}${key.hex()}"


def verify_password(stored: str, password: bytes) -> bool:
    _, iters, salt_hex, hash_hex = stored.split("$")
    salt = bytes.fromhex(salt_hex)
    expected = bytes.fromhex(hash_hex)
    derived = hashlib.pbkdf2_hmac(
        "sha256", password, salt, int(iters), dklen=len(expected)
    )
    return hmac.compare_digest(derived, expected)


if __name__ == "__main__":
    password = b"supersecret"
    start = time.perf_counter()
    stored = hash_password(password)
    elapsed_ms = (time.perf_counter() - start) * 1000
    print(f"PBKDF2 stored: {stored}")
    print(f"hash time: {elapsed_ms:.2f} ms")
    print(f"verify: {verify_password(stored, password)}")
